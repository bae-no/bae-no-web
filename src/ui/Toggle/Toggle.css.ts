import { style } from "@vanilla-extract/css";
import { sprinkles } from "../sprinkles.css";
import { theme } from "../tokens";

export const toggleRootStyle = style([
  sprinkles({
    minSize: "lg",
    bg: "black7",
    br: "md",
    p: "xxxs",
  }),
  style({
    width: "5rem",
    transition: "background-color 300ms",
    selectors: {
      '&[data-state="checked"]': {
        backgroundColor: theme.colors.orange2,
      },
    },
  }),
]);

export const toggleThumbStyle = style([
  sprinkles({
    display: "block",
    size: "lg",
    bg: "white",
    br: "half",
  }),
  {
    transition: "transform 300ms",
    selectors: {
      '&[data-state="checked"]': {
        transform: "translateX(calc(100% - 0.2rem))",
      },
    },
  },
]);
