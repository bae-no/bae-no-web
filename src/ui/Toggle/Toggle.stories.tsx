import { ComponentStory, ComponentMeta } from "@storybook/react";

import Toggle from "./Toggle";

export default {
  component: Toggle,
  title: "UI/Toggle",
} as ComponentMeta<typeof Toggle>;

const Template: ComponentStory<typeof Toggle> = (args) => <Toggle {...args} />;

export const Default = Template.bind({});
Default.args = {
  checked: true,
};
