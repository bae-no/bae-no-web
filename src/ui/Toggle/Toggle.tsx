import { Root, Thumb } from "@radix-ui/react-switch";
import clsx from "clsx";
import { ForwardedRef, forwardRef } from "react";
import { base } from "../reset.css";
import { toggleRootStyle, toggleThumbStyle } from "./Toggle.css";

interface ToggleProps {
  defaultChecked?: boolean;
  checked?: boolean;
  onCheckedChange?: (checked: boolean) => void;
  disabled?: boolean;
  required?: boolean;
  name?: string;
  value?: string;
}

function Toggle(props: ToggleProps, ref: ForwardedRef<HTMLButtonElement>) {
  return (
    <Root {...props} ref={ref} className={clsx([toggleRootStyle, base])}>
      <Thumb className={toggleThumbStyle} />
    </Root>
  );
}

export default forwardRef(Toggle);
