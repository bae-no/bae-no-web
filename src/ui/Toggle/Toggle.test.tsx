import { useState } from "react";

import { fireEvent, render } from "@testing-library/react";

import Toggle from "./Toggle";

const MockComponent = (props: unknown) => {
  const [checked, setChecked] = useState(false);
  return <Toggle {...props} checked={checked} onCheckedChange={setChecked} />;
}

describe("UI Toggle Component", () => {
  it("match snapshot", () => {
    const { container } = render(<Toggle />);
    expect(container).toMatchSnapshot();
  });

  it("should trigger onCheckedChange", () => {
    const mockClick = jest.fn();
    const { getByTestId } = render(
      <Toggle data-testid="toggle" onCheckedChange={mockClick} />
    );
    const target = getByTestId("toggle");

    fireEvent.click(target);
    expect(mockClick).toHaveBeenCalled();
    fireEvent.click(target);
    expect(mockClick).toHaveBeenCalledTimes(2);
  });

  it("should handle checked state as uncontrolled", () => {
    const { getByTestId } = render(<Toggle data-testid="toggle" />);
    const target = getByTestId("toggle");
    fireEvent.click(target);

    expect(target).toBeChecked();
    fireEvent.click(target);
    expect(target).not.toBeChecked();
  });

  it("should handle checked state as controlled", () => {
    const { getByTestId } = render(<MockComponent data-testid="toggle" />);
    const target = getByTestId("toggle");
    fireEvent.click(target);

    expect(target).toBeChecked();
    fireEvent.click(target);
    expect(target).not.toBeChecked();
  });

  it("should render correct default checked", () => {
    const { getByTestId } = render(
      <Toggle defaultChecked data-testid="toggle" />
    );
    const target = getByTestId("toggle");

    expect(target).toBeChecked();
    fireEvent.click(target);
    expect(target).not.toBeChecked();
  });

  it("should handle disabled state", () => {
    const mockClick = jest.fn();
    const { getByTestId } = render(
      <Toggle disabled data-testid="toggle" onCheckedChange={mockClick} />
    );
    const target = getByTestId("toggle");

    fireEvent.click(target);
    expect(target).toBeDisabled();
    expect(target).not.toBeChecked();
    expect(mockClick).not.toHaveBeenCalled();
  });
});
