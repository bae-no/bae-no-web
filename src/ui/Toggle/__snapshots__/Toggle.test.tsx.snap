// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UI Toggle Component match snapshot 1`] = `
<div>
  <button
    aria-checked="false"
    class="Toggle_toggleRootStyle__f6mp672 Toggle__f6mp670 sprinkles_padding_xxxs_mobile__1vna7gb23 sprinkles_borderRadius_md_mobile__1vna7gbir sprinkles_minWidth_lg_mobile__1vna7gbcl sprinkles_minHeight_lg_mobile__1vna7gbgc sprinkles_backgroundColor_black7_lightMode__1vna7gbnt Toggle_toggleRootStyle__f6mp671 reset_base__4op0mk0"
    data-state="unchecked"
    role="switch"
    type="button"
    value="on"
  >
    <span
      class="Toggle_toggleThumbStyle__f6mp674 Toggle__f6mp673 sprinkles_borderRadius_half_mobile__1vna7gbi3 sprinkles_width_lg_mobile__1vna7gbbc sprinkles_height_lg_mobile__1vna7gbf3 sprinkles_backgroundColor_white_lightMode__1vna7gbo1 sprinkles_display_block_mobile__1vna7gb6"
      data-state="unchecked"
    />
  </button>
</div>
`;
